File: backend/app/models/organization.py
from sqlalchemy import Column, Integer, String, DateTime, func, ForeignKey, UniqueConstraint
from sqlalchemy.orm import relationship
from backend.app.db.base import Base


class Organization(Base):
    __tablename__ = "organizations"
    
    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    memberships = relationship("Membership", back_populates="organization", cascade="all, delete-orphan")
    invitations = relationship("Invitation", back_populates="organization", cascade="all, delete-orphan")


class Membership(Base):
    __tablename__ = "memberships"
    
    id = Column(Integer, primary_key=True, index=True)
    org_id = Column(Integer, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False)
    user_id = Column(Integer, ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    role = Column(String(20), nullable=False)  # owner, admin, member
    
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    organization = relationship("Organization", back_populates="memberships")
    user = relationship("User", backref="memberships")
    
    # Unique constraint: user can only have one membership per org
    __table_args__ = (
        UniqueConstraint("org_id", "user_id", name="uq_membership_org_user"),
    )


class Invitation(Base):
    __tablename__ = "invitations"
    
    id = Column(Integer, primary_key=True, index=True)
    org_id = Column(Integer, ForeignKey("organizations.id", ondelete="CASCADE"), nullable=False)
    email = Column(String(255), nullable=False, index=True)
    role = Column(String(20), nullable=False)  # owner, admin, member
    token = Column(String(36), unique=True, nullable=False, index=True)  # UUID
    status = Column(String(20), nullable=False, default="pending")  # pending, accepted, expired
    expires_at = Column(DateTime(timezone=True), nullable=True)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    
    # Relationships
    organization = relationship("Organization", back_populates="invitations")

File: frontend/types/organization.ts
export interface Organization {
  id: number;
  name: string;
  created_at: string;
}

export interface Member {
  id: number;
  user_id: number;
  org_id: number;
  role: string;
  created_at: string;
  user_email: string;
  user_name: string;
}

export interface Invitation {
  id: number;
  org_id: number;
  email: string;
  role: string;
  token: string;
  status: string;
  expires_at: string | null;
  created_at: string;
}

export interface InviteRequest {
  email: string;
  role: 'owner' | 'admin' | 'member';
}

export interface UpdateMemberRoleRequest {
  role: 'owner' | 'admin' | 'member';
}

